### PR TITLE
chore: add project access resource

### PR DESCRIPTION
**Pull request by gastonfournier**
## About the changes
1. [Bump go server SDK dependency](https://github.com/Unleash/terraform-provider-unleash/pull/58/files#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6R6)
2. Adapt existing implementation to the new version: https://github.com/Unleash/terraform-provider-unleash/pull/58/files#diff-e1a140a60de11428c7904efeb759dfbf1f26df12e4b8710d228b3ef618df321cR129
3. Implement project access resources and tests: https://github.com/Unleash/terraform-provider-unleash/pull/58/files#diff-f2909944496491fce4b7d70ff2027e8355c5396b6e646367fbfe0dc444923084R21
4. Run tests against [specific docker images for OSS and Enterprise](https://github.com/Unleash/terraform-provider-unleash/pull/58/files#diff-faff1af3d8ff408964a57b2e475f69a6b7c7b71c9978cccc8f471798caac2c88R64-R66)
5. Add examples [d63f055](https://github.com/Unleash/terraform-provider-unleash/pull/58/commits/d63f055013678add0146fbcb59b8bcc406eb4c49)

### CODE REVIEW

File: internal/provider/project_access_resource_test.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+					resource "unleash_project_access" "sample_project_access" {
+						project = unleash_project.sample_project.id
+						roles = [
+							{
+								role = data.unleash_role.project_member_role.id
+								users = [
+									unleash_user.test_user.id
+								]
+								groups = []
```
Comment (gastonfournier) [inline internal/provider/project_access_resource_test.go:51]: Another important detail. We still can't create groups, but it's possible to add group access if you know the id of the group.

File: internal/provider/project_access_resource.go [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+		return
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
+	tflog.Debug(ctx, "Finished updating projectAccess data source", map[string]any{"success": true})
+}
+
+func (r *projectAccessResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	resp.Diagnostics.AddWarning("Resource Not Deleted", "The projectAccess resource was removed from the Terraform state, but not deleted from the actual system. This is to avoid potential mistakes. Instead of deleting projectAccess you may just delete the whole project")
+}
```
[the remark this answers is not in the document]
Comment (sighphyre) [reply internal/provider/project_access_resource.go:182]: Don't think I disagree but I can see a bunch of reasons why you wouldn't want to delete a project - like it has archived toggles, history or settings that you care about. 

I think stopping people from locking themselves out for now is the right call but we shouldn't probably stop that from being a thing on the Unleash side

File: internal/provider/project_access_resource.go [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+
+/** Helper methods **/
+
+func (r *projectAccessResource) upsertProjectAccess(ctx context.Context, plan projectAccessResourceModel) diag.Diagnostics {
```
Comment (sighphyre) [inline internal/provider/project_access_resource.go:186]: This one is scary! Don't think we can do anything about it now but there's a lot of things that can go wrong here and backtracking state when something goes belly up is going to be hard!

We can't do this on the unleash side as a bulk operation?

File: internal/provider/permission_data_source_test.go [hunk 4]
```diff
@@ -1,12 +1,16 @@
...
 
 import (
+	"os"
 	"testing"
 
 	"github.com/hashicorp/terraform-plugin-testing/helper/resource"
 )
 
 func TestAccPermissionDataSource(t *testing.T) {
+	if os.Getenv("UNLEASH_ENTERPRISE") == "false" {
```
Comment (sighphyre) [inline internal/provider/permission_data_source_test.go:11]: Only noticed this now and this isn't a problem but the enterprise tests are opt out? This check would pass if the env var is not set?

**Review by sighphyre — APPROVED** [review]
Woooo!

File: internal/provider/project_access_resource.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (gastonfournier) [reply internal/provider/project_access_resource.go:182]: Good point! 

The way I thought implementing deletion was by setting access to no one (`users: [], groups: []`) but it'd be a bit misleading: If you're removing the access resource from Terraform configuration, what would be your expectation?

That's still "doable" in Unleash, although now it's not possible from the UI ([projects should always have a user](https://unleash-internal.slack.com/archives/C046J17G6P7/p1697456576634689)). 

But errs to safety and opens up the opportunity to gather feedback.

File: internal/provider/project_access_resource.go (hunk 3 above)
Comment (gastonfournier) [reply internal/provider/project_access_resource.go:186]: It is, we just send the desired state and then Unleash server handles it in a transactional way: https://github.com/ivarconr/unleash-enterprise/blob/cb178bc30d1232ad5f3c83473e45960327d7af93/src/project-access/project-access-service.ts#L32-L48

File: internal/provider/permission_data_source_test.go (hunk 4 above)
Comment (gastonfournier) [reply internal/provider/permission_data_source_test.go]: Spot on! It should be `os.Getenv("UNLEASH_ENTERPRISE") != "true"`

Comment (gastonfournier) [reply internal/provider/permission_data_source_test.go]: [f2270a0](https://github.com/Unleash/terraform-provider-unleash/pull/58/commits/f2270a0e998ba1d8cad384311d790b55a6985eb5)

File: internal/provider/project_access_resource.go (hunk 3 above)
Comment (sighphyre) [reply internal/provider/project_access_resource.go:186]: I'm dumb, sorry, I realise this is one call just a large method

File: internal/provider/project_access_resource.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sighphyre) [reply internal/provider/project_access_resource.go:182]: > But errs to safety and opens up the opportunity to gather feedback.

You're never going to hear me disagree with a statement like this!

I think this is totally good enough for now!

**Pull request merged**